File: frontend/soa-frontend/src/app/tour/service/review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Review } from '../model/review';


@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  private apiUrl = 'http://localhost:8082/reviews';

  constructor(private http: HttpClient) {}

  getReviewsByTour(tourId: string): Observable<Review[]> {
    return this.http.get<Review[]>(`${this.apiUrl}/by-tour?tourId=${tourId}`);
  }

  addReview(review: Review): Observable<any> {
    const headers = new HttpHeaders({ 'Content-Type': 'application/json' });
    return this.http.post(this.apiUrl, review, { headers });
  }
}

File: backend/tours/service/ReviewService.go
package service

import (
	"context"
	"database-example/model"
	"database-example/repo"
)

type ReviewService struct {
	repo repo.ReviewRepo
}

func NewReviewService(r repo.ReviewRepo) *ReviewService {
	return &ReviewService{repo: r}
}

func (s *ReviewService) AddReview(ctx context.Context, review *model.Review) (*model.Review, error) {
	return s.repo.AddReview(ctx, review)
}

func (s *ReviewService) GetReviewsByTour(ctx context.Context, tourId string) (model.Reviews, error) {
	return s.repo.GetReviewsByTour(ctx, tourId)
}
